### PR TITLE
DHC Buried Pipes

### PR DESCRIPTION
**Pull request by bravache**
## Update 06/07:

I have added the model `PlugFlowDiscretize` to address some of the concerns from the previous model, namely: how can we justify a constant temperature boundary condition (constant along the length of the pipe), especially in cases with uninsulated pipes and for long stretch of pipes.

The main change is that this model allows to discretize `PlugFlowCore` axially. In some ways, this goes against the main benefit of the plug flow model, which is to avoid discretization of the problem, but on the other hand, it allows to more easily justify a semi-fixed temperature boundary for each segment.

I also toyed with the idea of formalizing a "virtual insulation layer" parameter in the model, but it is actually tricky since this layer would take on the soil thermal property, and the pipe is currently not coupled with the soil. This means that if I were to formalize this virtual insulation layer, I would have to also add the soil as a parameter to the `PlugFlowDiscretize`. In the end, I opted for not adding that option, but showcased how to use virtual insulation in the example `DiscretizedBuriedPipes` (by simply making the insulation take on the soil conductivity and making the insulation thickness as the virtual layer thickness).

Playing with this `PlugFlowDiscretize` model in the example confirmed what we suspected: for uninsulated pipes, the number of segment has a significant impact on heat transfer. This impact is not significant for insulated pipes. In the example below, I run 90degC water for 10km, either with no insulation (technically, there is an insulation layer with 100W/mK conductivity) or with 2cm of mineral wool (k=0.035W/mK).
image

With regard to virtual insulation, the observation is that it can offset the need for discretization to a certain extent. The soil higher conductivity (k=1.58W/mK) does not allow us to fully substitute actual insulation with a layer of soil. In the example below, I run 20degC water for 10km, with either 20cm or 0.2cm of virtual insulation. What we see is that when the pipe is segmented in 10 pieces, the virtual insulation thickness has a moderate impact on the outlet temperature, and when the pipe is not discretized, we see that thicker virtual insulation slightly reduces the gap between 1seg-10seg.
image

I have tried coming up with an analytical solution for the heat balance between the plug flow pipe model and the ground coupling model which could recommend the right number of segment or virtual insulation thickness, but was not successful. I am open to suggestion though.

---

This PR introduces two new models `GroundCoupling` and `UndisturbedSoilTemperature`, which are to be used in the modeling of heat transfer of buried pipes found in DHC.

`UndisturbedSoilTemperature` is a model of the boundary conditions for the prescribed underground temperature. The model relies on information about surface temperature (defined by a sinusoid) and about soil thermal properties, and effectively dampens and shift the sinusoid with increasing depth.

image

`GroundCoupling` is a model of heat transfer for a network of coaxial buried pipes. The model relies on the potential flow theory and the mirror image technique, as described in [Kusuda (1981)](https://www.govinfo.gov/content/pkg/GOVPUB-C13-1d38cb64ce5be577e23afda5081690ee/pdf/GOVPUB-C13-1d38cb64ce5be577e23afda5081690ee.pdf), which is the same base model eventually used for single and multiple pipes equations suggested in the ASHRAE District Heating/Cooling Guides. 

For each buried pipe (source), a heat sink is virtually created on the other side of the surface, which helps in bounding the heat transfer from the pipe.

image

For a single pipe, this results in a simple model that is equivalent to the cylindrical radial, steady-state heat transfer equation between the surface of the pipe and the mirror sink. For a multiple pipe network, each pipe transfers heat with other pipes and their sinks.

This model is extensively used in the literature, and should be validated with measurement data, especially for quantifying transient error.

Note: One thing I can't really wrap my head around is that the model equation suggests that the surface temperature (i.e. the mirror isotherm) is equal to the ground reference temperature, but multiple sources suggest to use the undisturbed soil temperature (which is a transformation of the surface temperature) as the reference temperature, even in conjonction with this model. I am unsure how to reconcile those two observations.

### CODE REVIEW

**Review by AntoineGautier — CHANGES_REQUESTED** [review]
Please see my comments.
A discussion would be useful for me to understand exactly the implication of the steady-state heat transfer computation.

**Review by bravache — COMMENTED** [review]
I addressed all of the smaller comments, but I still need to work on the assumptions and the validation comment.

File: Buildings/Fluid/BuriedPipes/Examples/DiscretizedBuriedPipe.mo
Comment (AntoineGautier) [inline Buildings/Fluid/BuriedPipes/Examples/DiscretizedBuriedPipe.mo:23]: It would be clearer to have graphical annotations for those declarations so that we can simply access the inner parameters.
This remark holds for all the examples.

File: Buildings/Fluid/BuriedPipes/Examples/TwoBuriedPipes.mo
Comment (AntoineGautier) [inline Buildings/Fluid/BuriedPipes/Examples/TwoBuriedPipes.mo:80]: Missing description string

File: Buildings/Fluid/BuriedPipes/Examples/TwoBuriedPipes.mo
Comment (AntoineGautier) [inline Buildings/Fluid/BuriedPipes/Examples/TwoBuriedPipes.mo:112]: Missing description string

File: Buildings/Examples/VAVReheat/Validation/BaseClasses/Floor.mo
Comment (AntoineGautier) [inline Buildings/Examples/VAVReheat/Validation/BaseClasses/Floor.mo:1]: This class should not be modified.

**Review by AntoineGautier — CHANGES_REQUESTED** [review]
Please see my inline comments, especially the one about the alternative approach with the computation of the pipe external surface temperature within `PlugFlowPipe`.

Looking at the translation statistics of `Buildings.Fluid.BuriedPipes.Examples.DiscretizedBuriedPipe` I don't think it is a valid option for modeling DHC systems.
`Sizes after manipulation of the nonlinear systems: {8, 40}`

- [ ] Also, please update the release notes in the top-level package.

File: Buildings/Fluid/BuriedPipes/Examples/DiscretizedBuriedPipe.mo
Comment (bravache) [reply Buildings/Fluid/BuriedPipes/Examples/DiscretizedBuriedPipe.mo]: Did you mean specifically for `soiDat` or also for other objects? I've added graphical annotation for `soiDat` for all examples.

File: Buildings/BoundaryConditions/GroundTemperature/UndisturbedSoilTemperature.mo
Comment (bravache) [reply Buildings/BoundaryConditions/GroundTemperature/UndisturbedSoilTemperature.mo:116]: The sentence I have in the documentation is currently almost verbatim copied from the guide:

> In the calculation of the climatic constants (...) the ground surface temperature is assumed to equal the air temperature, which is an acceptable assump-
tion for most design calculations.

The qualification you propose seems pretty specific to the example that they've chosen in the guide. I don't know if I could generalize it to say that all surface type (or even just concrete pavement) will consistently increase soil temperature in the summer.

**Review by bravache — COMMENTED** [review]
@AntoineGautier I addressed and resolved most of the rename, docs and graphics changes and I've opened some of the thread for discussion. I will address the comment about connecting the `PlugFlowPipe` and `GroundCoupling` heat equation tomorrow. I still have to think a bit more about it.

File: Buildings/Fluid/BuriedPipes/GroundCoupling.mo
Comment (bravache) [reply Buildings/Fluid/BuriedPipes/GroundCoupling.mo]: Just to confirm, the paper you wanted to link was [Dynamic equation-based thermo-hydraulic pipe model for district heating and cooling systems](https://www.sciencedirect.com/science/article/pii/S0196890417307975) (not sure how this happened but your link is just a few digits off, but amazingly, equation (3) in the paper you linked is somewhat relevant).

Regarding your observation, I am not sure I fully understand. In particular, when you say:
> In this equation ports[:,seg].T corresponds to the temperature at the exterior surface of the pipe, which is the variable the equation is solved for (ports[:,seg].Q_flow being computed by the instance of PlugFlowPipe this component is connected to).

I do not understand why `PlugFlowPipe` is solving for `Q` and `GroundCoupling` is solving for `T_b` (to reprise the notation from Van Der Heijde). My understanding is that the models don't solve for one or the other, but simply define the relationships between `T_b`, `Q` and other variables, some of which are known.
Put differently, if I were to flatten out the two models into one, I would pretty much end up with a system of three equations:
```
(1) T_out = T_b + (T_in - T_b) * E
(2) Q  = (T_out - T_in) * C
(3) Q = (T_b - T_g) * P
```
where `E`, `C` and `P` are constant terms (assuming constant mass flow rate). Equations (1) and (2) are from the `PlugFlowPipe` model and (3) is from  `GroundCoupling` (for a single pipe).
The way I understand your comment, I would just be shuffling the equations around, but in essence, would keep the same system:
```
(1) T_b = (T_out - T_in) * E / (1 - E)
(2) T_out = T_in + Q / C
(3) Q = (T_b - T_g) * P
```
Am I getting that right? If so, I don't fully understand why this would be interpreted differently par the solver.

Comment (AntoineGautier) [reply Buildings/Fluid/BuriedPipes/GroundCoupling.mo]: You are right to consider the system of equations in its whole and I realize that my comment is confusing.
Let me try again.

1. The "contract" behind the `PlugFlowPipe` model is that a temperature boundary condition can be provided that is decoupled from the fluid temperature at the time scale at which the integration is done, that is the transport delay along the pipe. Based on that, the model is able to provide an analytical formulation of the local heat flow rate, integrate it and compute the fluid outlet temperature. 
2. The `GroundCoupling` model is based on the average pipe-soil interface temperature. Based on that, the model provides an analytical formulation of the average heat flow rate.

It seems to me that coupling those two models yields an ambiguous computation scheme where the contract 1. is no more satisfied. But that contract is also no more needed since we are able to compute the average heat flow rate from 2. So what we really need 1. to provide is rather a formulation of the mean surface temperature, knowing the average heat flow rate and fluid temperature variation along the pipe.

File: Buildings/Fluid/BuriedPipes/Examples/DiscretizedBuriedPipe.mo
Comment (AntoineGautier) [reply Buildings/Fluid/BuriedPipes/Examples/DiscretizedBuriedPipe.mo]: I meant for all instances of any record class. We usually instantiate them by drag and drop (so with graphical annotations).

File: Buildings/Fluid/BuriedPipes/GroundCoupling.mo
Comment (bravache) [reply Buildings/Fluid/BuriedPipes/GroundCoupling.mo:263]: I am not quite sure how to answer your initial question. The initial simplification that goes into calculating the undisturbed ground temperature is that the surface temperature is an annual sinusoid. This simplification is going to be more or less accurate at smaller time scale depending on your climate, but typically would fail to precisely encompass daily variation.
As you go deeper into the ground, the daily temperature variation are less important, and the model would be more accurate on smaller time scale. For instance, depending on the soil properties, if you are 10m deep, you might expect the ground to be almost at a constant temperature year round, and the model would accurately represent that at even the second time-scale.

image

With that in mind, the statement about the heat transfer being steady state is actually meant to qualify the heat transfer between the pipes and the soil. The temporal scale at which this is relevant depends on the problem. If the pipe temperature vary significantly throughout the day and the user need to project hourly data, this might not be an accurate representation. On the other hand, if the fluid temperature is relatively stable, and the pipe is either well insulated or deep enough in the ground (so that the undisturbed ground temperature model is accurate at faster time scale), then the model might accurately represent steady-state heat losses.

What I meant by "seasonal heat storage is modeled in the ground temperature model" is probably poorly formulated, but is meant to convey that the undisturbed ground temperature is varying throughout the year, and that its calculation takes into account the thermal diffusivity of the soil.

**Review by bravache — COMMENTED** [review]
@AntoineGautier After spending some time trying to adjust the model, I think that this is as good as it might get. Once we get the data from OSU, we might make more discovery, but until then, I think I have exhausted all path for improving or evaluating it.
To summarize, this branch does not change the way heat is exchanged between a fluid and its surrounding. That problem is already modeled by `PlugFlowPipe` with sufficient accuracy and this is not improved here.
What this branch propose is a model of the undisturbed ground temperature and a model for the coupling between multiple pipes, following the ASHRAE District Heating and Cooling Guide recommendations. 

The undisturbed ground temperature model is an improvement as it allows to predict soil condition without measurements, as it only requires climatic conditions and soil properties. It might also be an improvement for capturing the actual temperature at burial depth when compared to a steady-state heat flow equation (which was essentially used in the `PlugFlowPipe` model validation), but this remains to be proven. The existing data from AIT is not sufficient to prove this.

The ground coupling model is an improvement mainly because it allows the coupling of multiple pipes, which would otherwise not be possible with just `PlugFlowPipe`. For a single pipe problem, the ground coupling model is similar to how the heat flow is defined in the `PlugFlowPipe` model, although it uses the undisturbed ground temperature and mirror images (which are `2d` away from the source, where `d` is the burial depth), as recommended by ASHRAE, instead of a known prescribed temperature.

There is still some details that need to be ironed out before this is merged, including adding a release note and potentially moving packages, but I would like to get a green light on whether or not we can move in with those models before finalizing all that.

File: Buildings/Fluid/BuriedPipes/package.mo
Comment (AntoineGautier) [reply Buildings/Fluid/BuriedPipes/package.mo:1]: Agreed.

File: Buildings/Fluid/BuriedPipes/GroundCoupling.mo
Comment (AntoineGautier) [reply Buildings/Fluid/BuriedPipes/GroundCoupling.mo]: If we envision to keep the discretized pipe model then the documentation of this model also needs to be updated to describe the axial discretization. And mention that the assumption of isothermic planes holds for each segment only.

File: Buildings/Fluid/BuriedPipes/package.order
Comment (AntoineGautier) [reply Buildings/Fluid/BuriedPipes/package.order:5]: I don't think `Buildings.Fluid.FixedResistances.Validation.PlugFlowPipes.PlugFlowAIT` takes into account the capacity of the ground as `thickness=0.0032 "Pipe wall thickness"` and `CPip=length*((dh + 2*thickness)^2 - dh^2)*Modelica.Constants.pi/4*cPip*rhoPip`. Only the thermal resistance seems to be corrected to account for the ground. In that respect the `GroundCoupling` model brings higher fidelity as if we were to simulate uninsulated burried pipes with the former approach, the outlet temperature would exhibit variations with the same dynamics as the air temperature.

File: Buildings/BoundaryConditions/GroundTemperature/UndisturbedSoilTemperature.mo
Comment (AntoineGautier) [reply Buildings/BoundaryConditions/GroundTemperature/UndisturbedSoilTemperature.mo]: I agree with your answer.

**Comment by AntoineGautier** [comment]
@bravache Thanks for the detailed explanations and the validation case. I agree to move forward with the formulation implemented here that avoids using the air temperature as a boundary condition for buried pipes. TODO before merging:

- [ ] See my last inline comments.
- [ ] Update the release notes in the top-level package.
- [ ] Maybe revisit the package structure, especially `BuriedPipes` should probably be relocated not directly under `Fluid`.
- [ ] Maybe sync with [IBPSA #1494](https://github.com/ibpsa/modelica-ibpsa/issues/1494).

This would make the branch ready to merge.
We will then need to further validate the pipe-to-pipe heat transfer with OSU data.

File: Buildings/Fluid/BuriedPipes/package.order
Comment (bravache) [reply Buildings/Fluid/BuriedPipes/package.order:5]: Yes, I noticed that as well. I think this is a mistake, either in how the validation model instantiate the `PlugFlowPipe` objects, or in how the `PlugFlowPipe` class is defined.
The model distinguishes between the pipe wall and the insulation, and the insulation heat capacity is ignored in the calculation of the equivalent volume. This makes sense in most application since the heat capacity of the insulation will likely be insignificant compared to the pipe wall, but this simplification is likely erroneous in cases where the surrounding soil is encompassed in the insulation.

File: Buildings/Examples/VAVReheat/Validation/BaseClasses/Floor.mo
Comment (bravache) [reply Buildings/Examples/VAVReheat/Validation/BaseClasses/Floor.mo:1]: Actually, it changed because the `master` branch file is encoded with BOM, so the formatting test fails. Not sure how this slipped through the cracks.

File: Buildings/Fluid/BuriedPipes/GroundCoupling.mo
Comment (bravache) [reply Buildings/Fluid/BuriedPipes/GroundCoupling.mo]: I updated the documentation to reflect the concept of axial segmentation and provide more information about the isothermic requirement.

File: Buildings/BoundaryConditions/GroundTemperature/UndisturbedSoilTemperature.mo
Comment (bravache) [reply Buildings/BoundaryConditions/GroundTemperature/UndisturbedSoilTemperature.mo]: I added a comment in the documentation that the soil diffusivity is assumed constant.

File: Buildings/BoundaryConditions/GroundTemperature/UndisturbedSoilTemperature.mo
Comment (bravache) [reply Buildings/BoundaryConditions/GroundTemperature/UndisturbedSoilTemperature.mo:47]: @AntoineGautier Are you ok with keeping the warning rather than blocking?

File: Buildings/Fluid/BuriedPipes/Examples/package.mo
Comment (bravache) [reply Buildings/Fluid/BuriedPipes/Examples/package.mo]: @AntoineGautier Looks good?

File: Buildings/BoundaryConditions/GroundTemperature/UndisturbedSoilTemperature.mo
Comment (bravache) [reply Buildings/BoundaryConditions/GroundTemperature/UndisturbedSoilTemperature.mo:116]: @AntoineGautier I've added a bit more documentation in this class and also in `Buildings.BoundaryConditions.GroundTemperature.BaseClasses.surfaceTemperature`. Let me know if you are satisfied.

File: Buildings/Fluid/BuriedPipes/package.order
Comment (bravache) [reply Buildings/Fluid/BuriedPipes/package.order]: @AntoineGautier Are you ok resolving this?

**Comment by bravache** [comment]
@AntoineGautier There are still a few comments that are "unresolved". Take a look and let me know if it's good to go.
Otherwise, I've reversed the changes to the `PlugFlowPipe`, and pushed them separately in an [IBPSA PR](https://github.com/ibpsa/modelica-ibpsa/pull/1503). Once that is merged, I will also need to update the example and validation models in this branch, but this can be done in a future PR, and this branch is fully operational regardless of the status if the IBPSA change.

One outstanding change is the proposed structure of the new components. In the end, I've opted to put the new `PlugFlowDiscretized` directly with the other existing pipe components in the `Fluid.FixedResistance` library. I also added two new library: `Buildings.BoundaryConditions.GroundTemperature` and `Buildings.Fluid.Geothermal.BuriedPipes`, which contain the undisturbed ground temperature module and the ground coupling model respectively. I think this structure makes the most sense, and we can punt the issue of having a dedicated Pipe library for a separate branch.

File: Buildings/Fluid/BuriedPipes/package.order
Comment (AntoineGautier) [reply Buildings/Fluid/BuriedPipes/package.order]: Yes

**Review by AntoineGautier — APPROVED** [review]

File: Buildings/Fluid/BuriedPipes/Examples/package.mo
Comment (AntoineGautier) [reply Buildings/Fluid/BuriedPipes/Examples/package.mo]: Yes

File: Buildings/BoundaryConditions/GroundTemperature/UndisturbedSoilTemperature.mo
Comment (AntoineGautier) [reply Buildings/BoundaryConditions/GroundTemperature/UndisturbedSoilTemperature.mo:47]: OK

**Comment by bravache** [comment]
@mwetter This is ready to be merged.

**Review by mwetter — APPROVED** [review]
This is ready to merge when all tests pass.

**Pull request merged**
